File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
      let mid = Math.floor((left + right) / 2);
      
      if (nums[mid] === target) {
          return mid;
      }

      // Check if the left half is sorted
      if (nums[left] <= nums[mid]) {
          // Check if the target is in the left half
          if (nums[left] <= target && target < nums[mid]) {
              right = mid - 1;
          } else {
              left = mid + 1;
          }
      // Check if the right half is sorted
      } else {
          // Check if the target is in the right half
          if (nums[mid] < target && target <= nums[right]) {
              left = mid + 1;
          } else {
              right = mid - 1;
          }
      }
  }
  return -1;
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        sorted_nums = sorted(nums)
        
        def custom_compare(target, n):
            if target == n:
                return 0
            elif target < n:
                return -1
            else:
                return 1
        
        result = None
        left, right = 0, len(sorted_nums) - 1
        
        while left <= right:
            mid = (left + right) // 2
            comparison = custom_compare(target, sorted_nums[mid])
            if comparison == 0:
                result = sorted_nums[mid]
                break
            elif comparison < 0:
                right = mid - 1
            else:
                left = mid + 1
        
        return nums.index(result) if result is not None else -1